rfc/0017-standard-feature-toggle.md

```diff
@@ -0,0 +1,140 @@
+
+
+## Motivation
+
+Currently there is no easy way for developers to implement feature toggles unless they are using the custom payload packets directly with the networking API which requires more work.
+
+The drawbacks of developers using non-standardized protocols is the impossibility to write a common server mod to disable some features in some mods. This aims to reduce work to multiple different parties.
+
+
+## Explanation
+
+### How would this work?
+
+The idea behind this RFC is for mod developers there would be an API for which mods can register a feature, for a given feature you can set a client value (from the client config), and query if the feature is enabled or not (by checking if the server allows it then it returns the client value, else it returns false).  
+Each features are identified by an ID.  
+Servers can send a packet changing the values of those features by ID.  
+When a client joins a server it would send a or multiple packet(s) describing which features exist on the client.
+
+Mods can set whether the server must explicitely authorize the features to enable it, or explicitely disallow the features to disable it.
+
+### Networking specification
+
+To standardize this feature toggle system it needs a standardized networking specification so anyone can re-implement it with a custom API for different modding platforms (Spigot, Sponge, Forge, etc).
+
+The networking specification is divided in multiple parts:
+ - networking types
+ - networking packets
+
+All types used are either defined in this RFC or are used in MC's own networking. `S->C` means Server to Client and `C->S` means Client to Server.
+
+In this specification, the type `string` without parenthesis describes an UTF-8 string with a maximum length of `32767`.
+
+### Array `A[]`
+
+An array holds a list of values of the specified type.
+It also contains the size `N` of the array.
+
+| Fields | Type     | Description                                 |
+|:------:|:--------:|:--------------------------------------------|
+| size   | [VarI32] | The size of the array.                      |
+| 0      | A        | The first element of the array if present.  |
+| 1      | A        | The second element of the array if present. |
+| ...    | A        | An element of the array if present.         |
+| N-1    | A        | The N-1 element of the array if present.    |
+
+#### FeatureList
+
+| Fields    | Type     | Description                                                     |
+|:---------:|:--------:|:---------------------------------------------------------------:|
+| namespace | string   | The namespace of the features. Can be considered the modid too. |
+| features  | string[] | The features inside the given namespace.                        |
+
+#### FeatureAuthorization
+
+Can be described as an enum with two possible values: `ALLOW` or `DENY`.
```

Comment:
One question people may ask it how would more granular feature sets be allowed or denied given a more boolean-like state shown here.